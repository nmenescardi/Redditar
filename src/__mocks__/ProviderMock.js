import { Provider } from 'react-redux';
import Store from '../store';

const defaultStore = Store();

const ProviderMock = (props) => {
  const store = props.store || defaultStore.store;
  return <Provider store={store}>{props.children}</Provider>;
};

export default ProviderMock;
