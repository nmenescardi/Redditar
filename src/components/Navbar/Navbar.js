import React from 'react';

const Navbar = () => {
  return (
    <section className="navbar">
      <h1>Redditar</h1>
    </section>
  );
};

export default Navbar;
