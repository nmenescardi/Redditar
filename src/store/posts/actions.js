import {
  POSTS_SET_DATA,
  POSTS_SET_LOADING,
  VISITED_ADD_POST,
  LAYOUT_TOGGLE_SIDEBAR,
  POSTS_DISMISS,
  POSTS_DISMISS_ALL,
  POSTS_SELECT,
} from './types';

import { GET_POSTS_URL } from '../../utils/config';

export function postSetData(posts) {
  return {
    type: POSTS_SET_DATA,
    payload: posts,
  };
}
export function postSetLoading(loadingFlag) {
  return {
    type: POSTS_SET_LOADING,
    payload: loadingFlag,
  };
}
export function visitedAddPost(postId) {
  return {
    type: VISITED_ADD_POST,
    payload: postId,
  };
}
export function layoutToggleSidebar() {
  return {
    type: LAYOUT_TOGGLE_SIDEBAR,
  };
}
export function postDismiss(postId) {
  return {
    type: POSTS_DISMISS,
    payload: postId,
  };
}
export function postDismissAll() {
  return {
    type: POSTS_DISMISS_ALL,
  };
}
export function selectPost(postId) {
  return {
    type: POSTS_SELECT,
    payload: postId,
  };
}

export const getPosts = () => (dispatch, getState) => {
  const { posts } = getState().posts;

  if (posts.length === 0) {
    dispatch(postSetLoading(true));

    fetch(GET_POSTS_URL)
      .then((request) => request.json())
      .then((data) => {
        dispatch(postSetData(data.data.children));
        dispatch(postSetLoading(false));
      });
  } else {
    dispatch(postSetData(posts));
  }
};
