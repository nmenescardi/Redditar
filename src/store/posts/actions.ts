import { Dispatch } from 'redux';
import { Action } from '../../enums';
import { Post, postID } from '../../types';
import { GET_POSTS_URL } from '../../utils/config';

export function postSetData(posts: Post[]) {
  return {
    type: Action.POSTS_SET_DATA,
    payload: posts,
  };
}
export function postSetLoading(loadingFlag: boolean) {
  return {
    type: Action.POSTS_SET_LOADING,
    payload: loadingFlag,
  };
}
export function visitedAddPost(postId: postID) {
  return {
    type: Action.VISITED_ADD_POST,
    payload: postId,
  };
}
export function layoutToggleSidebar() {
  return {
    type: Action.LAYOUT_TOGGLE_SIDEBAR,
  };
}
export function postDismiss(postIds: postID[] | postID) {
  return {
    type: Action.POSTS_DISMISS,
    payload: postIds instanceof Array ? postIds : [postIds],
  };
}
export function postDismissAll() {
  return {
    type: Action.POSTS_DISMISS_ALL,
  };
}
export function selectPost(postId: postID) {
  return {
    type: Action.POSTS_SELECT,
    payload: postId,
  };
}

export const getPosts = () => (dispatch: Dispatch) => {
  dispatch(postSetLoading(true));

  fetch(GET_POSTS_URL)
    .then((request) => request.json())
    .then((data) => {
      dispatch(postSetData(data.data.children));
      dispatch(postSetLoading(false));
    });
};
