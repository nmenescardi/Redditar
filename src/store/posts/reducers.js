import {
  POSTS_SET_DATA,
  POSTS_SET_LOADING,
  VISITED_ADD_POST,
  LAYOUT_TOGGLE_SIDEBAR,
  POSTS_DISMISS,
} from './types';

const initialState = {
  posts: [],
  visited: [],
  dismissed: [],
  loading: false,
  sidebarOpen: true,
};

export function postReducer(state = initialState, action) {
  switch (action.type) {
    case POSTS_SET_DATA:
      return {
        ...state,
        posts: [...action.payload],
      };

    case POSTS_SET_LOADING:
      return {
        ...state,
        loading: action.payload,
      };

    case VISITED_ADD_POST:
      return {
        ...state,
        // Make it unique
        visited: [...state.visited, action.payload].filter(
          (v, i, a) => a.indexOf(v) === i
        ),
      };

    case POSTS_DISMISS:
      console.log(`action.payload`, action.payload);
      return {
        ...state,
        // Make it unique
        dismissed: [...state.dismissed, action.payload].filter(
          (v, i, a) => a.indexOf(v) === i
        ),
      };

    case LAYOUT_TOGGLE_SIDEBAR:
      return {
        ...state,
        sidebarOpen: !state.sidebarOpen,
      };

    default:
      return state;
  }
}
