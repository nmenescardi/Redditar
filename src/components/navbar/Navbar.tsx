import React from 'react';
import { connect } from 'react-redux';
import { layoutToggleSidebar } from '../../store/posts/actions';

type Props = {
  layoutToggleSidebar: () => { type: string };
};

const Navbar: React.FC<Props> = ({ layoutToggleSidebar }) => {
  return (
    <section className="navbar" onClick={layoutToggleSidebar}>
      <div className="navbar__menu">
        <div className="navbar__bar"></div>
        <div className="navbar__bar"></div>
        <div className="navbar__bar"></div>
      </div>
      <h1 className="navbar__title" data-testid="h1">
        Redditar
      </h1>
    </section>
  );
};

const mapDispatchToProps = {
  layoutToggleSidebar,
};
export default connect(null, mapDispatchToProps)(Navbar);
