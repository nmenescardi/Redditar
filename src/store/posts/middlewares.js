import { POSTS_DISMISS_ALL } from './types';
import { postDismiss } from './actions';

export const dimissAllMiddleware = (store) => (next) => (action) => {
  const state = store.getState();

  if (POSTS_DISMISS_ALL === action.type) {
    const allPostsId = state.posts.posts.map((post) => post.data.id);
    store.dispatch(postDismiss(allPostsId));
  }

  next(action);
};
