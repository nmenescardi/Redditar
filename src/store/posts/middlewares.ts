import { Middleware } from 'redux';
import { POSTS_DISMISS_ALL } from './types';
import { postDismiss } from './actions';
import { Post } from '../../types';

export const dismissAllMiddleware: Middleware =
  (store) => (next) => (action) => {
    const state = store.getState();

    if (POSTS_DISMISS_ALL === action.type) {
      const allPostsId = state.posts.posts.map((post: Post) => post.data.id);
      store.dispatch(postDismiss(allPostsId));
    }

    next(action);
  };
