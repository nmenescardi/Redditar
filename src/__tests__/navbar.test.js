/**
 * @jest-environment jsdom
 */

import { render } from '@testing-library/react';
import Navbar from '../components/navbar/Navbar';
import ProviderMock from '../__mocks__/ProviderMock';
import { layoutToggleSidebar } from '../store/posts/actions';
import { create } from 'react-test-renderer';

describe('testing the sidebar', () => {
  const getElements = () => {
    const sidebar = render(
      <ProviderMock>
        <Navbar layoutToggleSidebar={layoutToggleSidebar}></Navbar>
      </ProviderMock>
    );

    return { sidebar: sidebar, container: sidebar.container };
  };

  it('should have the section element', () => {
    const { container } = getElements();
    const mainSection = container.getElementsByClassName('navbar');
    expect(mainSection.length).toBe(1);
  });

  it('should toggle navbar visibility on main section click.', async () => {
    const { sidebar } = getElements();

    const mainSection = await sidebar.findByTestId('navbar');
    mainSection.click();

    // TODO: check the store to see the changes
  });
});

describe('<Navbar /> Snapshot', () => {
  test('Navbar UI rendering correctly', () => {
    const navbar = create(
      <ProviderMock>
        <Navbar />
      </ProviderMock>
    );
    expect(navbar.toJSON()).toMatchSnapshot();
  });
});
