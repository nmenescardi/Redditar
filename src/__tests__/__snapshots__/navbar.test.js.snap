// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> Snapshot Navbar UI rendering correctly 1`] = `
<section
  className="navbar"
  data-testid="navbar"
  onClick={[Function]}
>
  <div
    className="navbar__menu"
  >
    <div
      className="navbar__bar"
    />
    <div
      className="navbar__bar"
    />
    <div
      className="navbar__bar"
    />
  </div>
  <h1
    className="navbar__title"
    data-testid="h1"
  >
    Redditar
  </h1>
</section>
`;
